src/dataflow/src/render/reduce.rs

```diff
@@ -31,23 +76,427 @@ use super::context::Context;
+    aggr_funcs: Vec<AggregateFunc>,
+    /// Set of "skips" or calls to `nth()` an iterator needs to do over
+    /// the input to extract the relevant datums.
+    skips: Vec<usize>,
+    // The number of buckets in each layer of the reduction tree. Should
+    // be decreasing, and ideally, a power of two so that we can easily
+    // distribute values to buckets with `value.hashed() % bucketes[layer]`.
+    buckets: Vec<u64>,
+}
+
+/// Plan for computing a set of basic aggregations.
+///
+/// There's much less complexity when rendering basic aggregations.
+/// Each aggregation corresponds to one Differential reduce operator.
+/// That's it. However, we still want to present one final arrangement
+/// so basic aggregations present results with the same interface
+/// (one arrangement containing a row with all results) that accumulable
+/// and hierarchical aggregations do. To provide that, we render an
+/// additional reduce operator whenever we have multiple reduce aggregates
+/// to combine and present results in the appropriate order. If we
+/// were only asked to compute a single aggregation, we can skip
+/// that step and return the arragement provided by computing the aggregation
+/// directly.
+#[derive(Clone, Debug)]
+enum BasicPlan {
+    /// Plan for rendering a single basic aggregation. Here, the
+    /// first element denotes the index in the set of inputs
+    /// that we are aggregating over.
+    Single(usize, AggregateExpr),
+    /// Plan for rendering multiple basic aggregations.
+    /// These need to then be collated together in an additional
+    /// reduction. Each element represents the:
+    /// `(index of the set of the input we are aggregating over,
+    ///   the aggregation function)`
+    Multiple(Vec<(usize, AggregateExpr)>),
+}
+
+/// Plan for collating the results of computing multiple aggregation
+/// types.
+///
+/// TODO: could we express this as a delta join
+#[derive(Clone, Debug, Default)]
+struct CollationPlan {
+    /// Accumulable aggregation results to collate, if any.
+    accumulable: Option<AccumulablePlan>,
+    /// Hierarchical aggregation results to collate, if any.
+    hierarchical: Option<HierarchicalPlan>,
+    /// Basic aggregation results to collate, if any.
+    basic: Option<BasicPlan>,
+    /// When we get results back from each of the different
+    /// aggregation types, they will be subsequences of
+    /// the sequence aggregations in the original reduce expression.
+    /// We keep a map from output position -> reduction type
+    /// to easily merge results back into the requested order.
+    aggregate_types: Vec<ReductionType>,
+}
+
+impl ReducePlan {
+    /// Generate a plan for computing the specified type of aggregations.
+    ///
+    /// This function assumes that all of the supplied aggregates are
+    /// actually of the correct reduction type, and are a subsequence
+    /// of the total list of requested aggregations.
+    fn create_inner(
```

Comment:
Minor nits: `create_inner` sounds weird. Is it not just `create_from`? What does `_inner` refer to? Also, `expected_group_size` is probably ultimately a per-aggregation quantity, though I bet that's another issue entirely; using one value to upper bound all is certainly fine for now.

Reply:
NB: the `expected_group_size` thing is only a comment now because it is clear that it shows up as an explicit argument, whereas before it was just some random line in the code, so that's a win for clarity.